File: src/tracker.ts
import type { CostRecord, CostSink, UsageType } from './types.js';
import { MemorySink, FileSink, SupabaseSink, type SupabaseSinkOptions } from './sinks.js';
import { captchaPrice, smsPrice, proxyCostForBytes, llmCost, computeCost, PRICES } from './pricing.js';

export interface CostTrackerOptions {
  agent_id: string;
  reference_id?: string;             // e.g. ACTION_LOG_ID for weles
  sink?: 'memory' | 'file' | 'supabase';
  filePath?: string;                 // when sink === 'file'
  supabase?: SupabaseSinkOptions;    // when sink === 'supabase'
  /** Auto-flush hooks: register beforeExit/SIGINT/SIGTERM listeners that
   *  flush pending records before the process exits. Default: true. */
  autoFlush?: boolean;
}

export class CostTracker {
  private sink: CostSink;
  private buffer: CostRecord[] = [];
  private agent_id: string;
  private reference_id?: string;
  private flushed = false;

  constructor(opts: CostTrackerOptions) {
    this.agent_id = opts.agent_id;
    this.reference_id = opts.reference_id;
    if (opts.sink === 'supabase') {
      if (!opts.supabase) throw new Error('CostTracker: sink=supabase requires opts.supabase');
      this.sink = new SupabaseSink(opts.supabase);
    } else if (opts.sink === 'file') {
      if (!opts.filePath) throw new Error('CostTracker: sink=file requires opts.filePath');
      this.sink = new FileSink(opts.filePath);
    } else {
      this.sink = new MemorySink();
    }
    if (opts.autoFlush !== false) this.installExitHooks();
  }

  private installExitHooks(): void {
    const flush = () => { void this.flush().catch(() => {}); };
    process.on('beforeExit', flush);
    process.on('SIGINT', () => { flush(); setTimeout(() => process.exit(130), 100); });
    process.on('SIGTERM', () => { flush(); setTimeout(() => process.exit(143), 100); });
  }

  /** Append a raw record. Computes cost from pricing if cost_usd absent. */
  record(rec: Omit<CostRecord, 'cost_usd'> & { cost_usd?: number }): CostRecord {
    const cost_usd = typeof rec.cost_usd === 'number' ? rec.cost_usd : 0;
    const full: CostRecord = {
      service: rec.service,
      resource: rec.resource,
      usage_type: rec.usage_type,
      usage_amount: rec.usage_amount,
      cost_usd: round4(cost_usd),
      reference_id: rec.reference_id ?? this.reference_id,
      metadata: rec.metadata ?? {},
      created_at: rec.created_at ?? new Date().toISOString(),
    };
    this.buffer.push(full);
    return full;
  }

  recordCaptcha(service: string, taskType: string, override?: number): CostRecord {
    return this.record({
      service: `captcha_${service}`,
      resource: taskType,
      usage_type: 'solves',
      usage_amount: 1,
      cost_usd: override ?? captchaPrice(service, taskType),
    });
  }

  recordSms(provider: string, platform: string, override?: number): CostRecord {
    return this.record({
      service: `sms_${provider}`,
      resource: platform,
      usage_type: 'units',
      usage_amount: 1,
      cost_usd: override ?? smsPrice(provider, platform),
    });
  }

  recordProxyBytes(provider: string, bytes: number, isMobile = false): CostRecord {
    const key = isMobile && provider === 'oxylabs' ? 'oxylabs_mobile' : provider;
    return this.record({
      service: `proxy_${key}`,
      resource: isMobile ? 'mobile' : 'residential',
      usage_type: 'bytes',
      usage_amount: bytes,
      cost_usd: proxyCostForBytes(provider, bytes, isMobile),
    });
  }

  recordLlm(model: string, inputTokens: number, outputTokens: number, override?: number): CostRecord {
    return this.record({
      service: `llm_${normalizeLlmService(model)}`,
      resource: model,
      usage_type: 'tokens',
      usage_amount: inputTokens + outputTokens,
      cost_usd: override ?? llmCost(model, inputTokens, outputTokens),
      metadata: { input_tokens: inputTokens, output_tokens: outputTokens },
    });
  }

  recordCompute(instanceType: string, seconds: number, override?: number): CostRecord {
    return this.record({
      service: `compute_${instanceType.split('_')[0] ?? 'other'}`,
      resource: instanceType,
      usage_type: 'seconds',
      usage_amount: seconds,
      cost_usd: override ?? computeCost(instanceType, seconds),
    });
  }

  recordEmail(provider: string, count = 1, override?: number): CostRecord {
    const unit = (PRICES.email as Record<string, number>)[provider] ?? PRICES.email.default;
    return this.record({
      service: `email_${provider}`,
      resource: provider,
      usage_type: 'emails',
      usage_amount: count,
      cost_usd: override ?? unit * count,
    });
  }

  /** Total $ across all buffered records. */
  total(): number { return round4(this.buffer.reduce((a, r) => a + r.cost_usd, 0)); }

  /** Per-service cost rollup compatible with weles' service_costs JSONB. */
  snapshot(): { cost_usd: number; service_costs: Record<string, number>; records: CostRecord[] } {
    const service_costs: Record<string, number> = {};
    for (const r of this.buffer) service_costs[r.service] = round4((service_costs[r.service] ?? 0) + r.cost_usd);
    return { cost_usd: this.total(), service_costs, records: this.buffer.slice() };
  }

  /** Flush buffered records to the sink. Idempotent — safe to call multiple
   *  times; second call is a no-op. */
  async flush(): Promise<void> {
    if (this.flushed) return;
    if (this.buffer.length === 0) { this.flushed = true; return; }
    const stamped = this.buffer.map(r => ({ ...r, agent_id: this.agent_id }));
    await this.sink.write(stamped as CostRecord[]);
    this.flushed = true;
  }

  /** Underlying sink, for advanced usage (e.g. read-back during the same run). */
  getSink(): CostSink { return this.sink; }
}

function round4(n: number): number { return Math.round(n * 10000) / 10000; }

function normalizeLlmService(model: string): string {
  const m = model.toLowerCase();
  if (m.includes('gemini')) return 'gemini';
  if (m.includes('claude')) return 'claude';
  if (m.includes('gpt')) return 'openai';
  return 'other';
}

File: py/src/wisent_cost_tracker/tracker.py
"""CostTracker — Python mirror of @wisent/cost-tracker."""

import atexit
from dataclasses import dataclass
import math
import signal
from pathlib import Path
from typing import Any, Dict, List, Optional

from .pricing import (
    captcha_price,
    compute_cost,
    llm_cost,
    proxy_cost_for_bytes,
    sms_price,
    PRICES,
)
from .sinks import CostSink, FileSink, MemorySink, SupabaseSink
from .types import CostRecord

_USAGE_TYPES = {"solves", "tokens", "bytes", "seconds", "units", "emails"}


@dataclass
class CostTrackerOptions:
    agent_id: str
    reference_id: Optional[str] = None
    sink: str = "memory"            # "memory" | "file" | "supabase"
    file_path: Optional[Path] = None
    supabase_url: Optional[str] = None
    supabase_key: Optional[str] = None
    auto_flush: bool = True


def _round4(n: float) -> float:
    return round(n, 4)


def _normalize_llm_service(model: str) -> str:
    m = model.lower()
    if "gemini" in m:
        return "gemini"
    if "claude" in m:
        return "claude"
    if "gpt" in m:
        return "openai"
    return "other"


class CostTracker:
    def __init__(self, opts: CostTrackerOptions | None = None, **kwargs: Any) -> None:
        if opts is None:
            opts = CostTrackerOptions(**kwargs)
        if not isinstance(opts.agent_id, str) or not opts.agent_id.strip():
            raise ValueError("CostTracker: agent_id must be a non-empty string")
        self._opts = opts
        self._buffer: List[CostRecord] = []
        self._flushed = False

        if opts.sink == "supabase":
            if not (opts.supabase_url and opts.supabase_key):
                raise ValueError("CostTracker: sink='supabase' requires supabase_url + supabase_key")
            self._sink: CostSink = SupabaseSink(opts.supabase_url, opts.supabase_key)
        elif opts.sink == "file":
            if not opts.file_path:
                raise ValueError("CostTracker: sink='file' requires file_path")
            self._sink = FileSink(opts.file_path)
        else:
            self._sink = MemorySink()

        if opts.auto_flush:
            atexit.register(self._flush_sync)
            for sig in (signal.SIGINT, signal.SIGTERM):
                try:
                    signal.signal(sig, self._signal_flush)
                except (ValueError, OSError):
                    # signal may not be settable in non-main threads
                    pass

    def _signal_flush(self, _signum: int, _frame: Any) -> None:
        self._flush_sync()

    def _flush_sync(self) -> None:
        try:
            self.flush()
        except Exception:  # noqa: BLE001
            pass

    def record(
        self,
        service: str,
        usage_type: str,
        usage_amount: float,
        cost_usd: float,
        resource: Optional[str] = None,
        reference_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> CostRecord:
        if not isinstance(service, str) or not service.strip():
            raise ValueError("CostTracker: service must be a non-empty string")
        if usage_type not in _USAGE_TYPES:
            raise ValueError(f"CostTracker: unsupported usage_type {usage_type!r}")
        if type(usage_amount) not in (int, float) or not math.isfinite(usage_amount) or usage_amount <= 0:
            raise ValueError("CostTracker: usage_amount must be a finite number greater than zero")
        if type(cost_usd) not in (int, float) or not math.isfinite(cost_usd) or cost_usd < 0:
            raise ValueError("CostTracker: cost_usd must be a finite non-negative number")
        rec = CostRecord(
            service=service,
            resource=resource,
            usage_type=usage_type,  # type: ignore[arg-type]
            usage_amount=float(usage_amount),
            cost_usd=_round4(cost_usd),
            reference_id=reference_id or self._opts.reference_id,
            metadata=metadata or {},
            created_at=datetime.utcnow().isoformat() + "Z",
            agent_id=self._opts.agent_id,
        )
        self._buffer.append(rec)
        return rec

    def record_captcha(self, service: str, task_type: str, override: Optional[float] = None) -> CostRecord:
        return self.record(
            service=f"captcha_{service}",
            resource=task_type,
            usage_type="solves",
            usage_amount=1,
            cost_usd=override if override is not None else captcha_price(service, task_type),
        )

    def record_sms(self, provider: str, platform: str, override: Optional[float] = None) -> CostRecord:
        return self.record(
            service=f"sms_{provider}",
            resource=platform,
            usage_type="units",
            usage_amount=1,
            cost_usd=override if override is not None else sms_price(provider, platform),
        )

    def record_proxy_bytes(self, provider: str, num_bytes: int, is_mobile: bool = False) -> CostRecord:
        key = "oxylabs_mobile" if (is_mobile and provider == "oxylabs") else provider
        return self.record(
            service=f"proxy_{key}",
            resource="mobile" if is_mobile else "residential",
            usage_type="bytes",
            usage_amount=num_bytes,
            cost_usd=proxy_cost_for_bytes(provider, num_bytes, is_mobile),
        )

    def record_llm(
        self,
        model: str,
        input_tokens: int = 0,
        output_tokens: int = 0,
        override: Optional[float] = None,
        skill_id: Optional[str] = None,
    ) -> CostRecord:
        meta: Dict[str, Any] = {"input_tokens": input_tokens, "output_tokens": output_tokens}
        if skill_id:
            meta["skill_id"] = skill_id
        return self.record(
            service=f"llm_{_normalize_llm_service(model)}",
            resource=model,
            usage_type="tokens",
            usage_amount=input_tokens + output_tokens,
            cost_usd=override if override is not None else llm_cost(model, input_tokens, output_tokens),
            metadata=meta,
        )

    def record_compute(self, instance_type: str, seconds: float, override: Optional[float] = None) -> CostRecord:
        return self.record(
            service=f"compute_{instance_type.split('_')[0]}",
            resource=instance_type,
            usage_type="seconds",
            usage_amount=seconds,
            cost_usd=override if override is not None else compute_cost(instance_type, seconds),
        )

    def record_email(self, provider: str, count: int = 1, override: Optional[float] = None) -> CostRecord:
        unit = PRICES["email"].get(provider, PRICES["email"]["default"])
        return self.record(
            service=f"email_{provider}",
            resource=provider,
            usage_type="emails",
            usage_amount=count,
            cost_usd=override if override is not None else unit * count,
        )

    def total(self) -> float:
        return _round4(sum(r.cost_usd for r in self._buffer))

    def snapshot(self) -> Dict[str, Any]:
        service_costs: Dict[str, float] = {}
        for r in self._buffer:
            service_costs[r.service] = _round4(service_costs.get(r.service, 0) + r.cost_usd)
        return {
            "cost_usd": self.total(),
            "service_costs": service_costs,
            "records": [r.to_json() for r in self._buffer],
        }

    def flush(self) -> None:
        if self._flushed or not self._buffer:
            self._flushed = True
            return
        self._sink.write(self._buffer)
        try:
            from .onboarding import observe_accepted_usage

            observe_accepted_usage(self._buffer)
        except (OSError, TypeError, ValueError):
            # First-use telemetry must never alter an accepted sink write.
            pass
        self._flushed = True

    def get_sink(self) -> CostSink:
        return self._sink
